File: examples/blue-green/operator.ts
import * as k8s from "@pulumi/kubernetes";
import * as pulumi from "@pulumi/pulumi";

// Create a ServiceAccount.
interface ServiceAccountArgs {
    namespace: pulumi.Input<string>;
    provider: k8s.Provider;
}
export function createServiceAccount(
    name: string,
    args: ServiceAccountArgs,
): k8s.core.v1.ServiceAccount {
    return new k8s.core.v1.ServiceAccount(name, {
        metadata: {namespace: args.namespace},
    },{provider: args.provider});
}

// Create a Role.
interface RoleArgs {
    namespace: pulumi.Input<string>;
    provider: k8s.Provider;
}
export function createRole(
    name: string,
    args: RoleArgs,
): k8s.rbac.v1.Role {
    return new k8s.rbac.v1.Role(name, {
        metadata: {namespace: args.namespace},
        rules: [
            {
                apiGroups: [""],
                resources: [
                    "pods",
                    "services",
                    "services/finalizers",
                    "endpoints",
                    "persistentvolumeclaims",
                    "events",
                    "configmaps",
                    "secrets",
                ],
                verbs: [
                    "create",
                    "delete",
                    "get",
                    "list",
                    "patch",
                    "update",
                    "watch",
                ],
            },
            {
                apiGroups: ["apps"],
                resources: [
                    "deployments",
                    "daemonsets",
                    "replicasets",
                    "statefulsets",
                ],
                verbs: [
                    "create",
                    "delete",
                    "get",
                    "list",
                    "patch",
                    "update",
                    "watch",
                ],
            },
            {
                apiGroups: ["monitoring.coreos.com"],
                resources: ["servicemonitors"],
                verbs: [
                    "create",
                    "get",
                ],
            },
            {
                apiGroups: ["apps"],
                resourceNames: ["pulumi-kubernetes-operator"],
                resources: ["deployments/finalizers"],
                verbs: ["update"],
            },
            {
                apiGroups: [""],
                resources: ["pods"],
                verbs: ["get"],
            },
            {
                apiGroups: ["apps"],
                resources: [
                    "replicasets",
                    "deployments",
                ],
                verbs: ["get"],
            },
            {
                apiGroups: ["pulumi.com"],
                resources: ["*"],
                verbs: [
                    "create",
                    "delete",
                    "get",
                    "list",
                    "patch",
                    "update",
                    "watch",
                ],
            },
        ],
    },{provider: args.provider});
}

// Create a RoleBinding.
interface RoleBindingArgs {
    serviceAccountName: pulumi.Input<string>;
    roleName: pulumi.Input<string>;
    namespace: pulumi.Input<string>;
    provider: k8s.Provider;
}
export function createRoleBinding(
    name: string,
    args: RoleBindingArgs,
): k8s.rbac.v1.RoleBinding {
    return new k8s.rbac.v1.RoleBinding(name, {
        metadata: {namespace: args.namespace},
        subjects: [{
            kind: "ServiceAccount",
            name: args.serviceAccountName,
            namespace: args.namespace,
        }],
        roleRef: {
            kind: "Role",
            name: args.roleName,
            apiGroup: "rbac.authorization.k8s.io",
        },
    },{provider: args.provider});
}

export interface PulumiKubernetesOperatorArgs {
    namespace: pulumi.Input<string>;
    provider: k8s.Provider,
}
export class PulumiKubernetesOperator extends pulumi.ComponentResource {
    public readonly deployment: k8s.apps.v1.Deployment;
    constructor(name: string,
        args: PulumiKubernetesOperatorArgs,
        opts: pulumi.ComponentResourceOptions = {}) {
        super("pulumi-kubernetes-operator", name, args, opts);

        // Create the service account.
        const serviceAccount = createServiceAccount(name, {
            namespace: args.namespace,
            provider: args.provider,
        });
        const serviceAccountName = serviceAccount.metadata.name;

        // Create RBAC role and binding.
        const role = createRole(name, {
            namespace: args.namespace,
            provider: args.provider,
        });
        const roleName = role.metadata.name;
        const roleBinding = createRoleBinding(name, {
            serviceAccountName: serviceAccountName,
            roleName: roleName,
            namespace: args.namespace,
            provider: args.provider,
        });

        // Create the deployment.
        this.deployment = new k8s.apps.v1.Deployment(name, {
            metadata: {namespace: args.namespace},
            spec: {
                replicas: 1,
                selector: {
                    matchLabels: {
                        name: "pulumi-kubernetes-operator",
                    },
                },
                template: {
                    metadata: {
                        labels: {
                            name: "pulumi-kubernetes-operator",
                        },
                    },
                    spec: {
                        serviceAccountName: serviceAccountName,
                        imagePullSecrets: [{
                            name: "pulumi-kubernetes-operator",
                        }],
                        containers: [{
                            name: "pulumi-kubernetes-operator",
                            image: "pulumi/pulumi-kubernetes-operator:v0.0.5",
                            command: ["pulumi-kubernetes-operator"],
                            args: ["--zap-level=debug"],
                            imagePullPolicy: "Always",
                            env: [
                                {
                                    name: "WATCH_NAMESPACE",
                                    valueFrom: {
                                        fieldRef: {
                                            fieldPath: "metadata.namespace",
                                        },
                                    },
                                },
                                {
                                    name: "POD_NAME",
                                    valueFrom: {
                                        fieldRef: {
                                            fieldPath: "metadata.name",
                                        },
                                    },
                                },
                                {
                                    name: "OPERATOR_NAME",
                                    value: "pulumi-kubernetes-operator",
                                },
                            ],
                        }],
                    },
                },
            },
        }, {dependsOn: roleBinding});
    }
}

File: lib/python/pulumi_crds/pulumi/v1alpha1/operator.py
import pulumi
import pulumi_kubernetes as kubernetes


class Operator(pulumi.ComponentResource):
    def __init__(self, name, opts=None):
        super().__init__('pulumi:v1alpha1:Operator', name, None, opts)

        self.operator_service_account = kubernetes.core.v1.ServiceAccount("operatorServiceAccount", metadata={
            "name": "pulumi-kubernetes-operator",
        }, opts=pulumi.ResourceOptions(parent=self))

        self.operator_role = kubernetes.rbac.v1.Role("operatorRole",
                                                metadata={
                                                    "name": "pulumi-kubernetes-operator",
                                                },
                                                rules=[
                                                    {
                                                        "api_groups": [""],
                                                        "resources": [
                                                            "pods",
                                                            "services",
                                                            "services/finalizers",
                                                            "endpoints",
                                                            "persistentvolumeclaims",
                                                            "events",
                                                            "configmaps",
                                                            "secrets",
                                                        ],
                                                        "verbs": [
                                                            "create",
                                                            "delete",
                                                            "get",
                                                            "list",
                                                            "patch",
                                                            "update",
                                                            "watch",
                                                        ],
                                                    },
                                                    {
                                                        "api_groups": ["apps"],
                                                        "resources": [
                                                            "deployments",
                                                            "daemonsets",
                                                            "replicasets",
                                                            "statefulsets",
                                                        ],
                                                        "verbs": [
                                                            "create",
                                                            "delete",
                                                            "get",
                                                            "list",
                                                            "patch",
                                                            "update",
                                                            "watch",
                                                        ],
                                                    },
                                                    {
                                                        "api_groups": ["monitoring.coreos.com"],
                                                        "resources": ["servicemonitors"],
                                                        "verbs": [
                                                            "create",
                                                            "get",
                                                        ],
                                                    },
                                                    {
                                                        "api_groups": ["apps"],
                                                        "resource_names": ["pulumi-kubernetes-operator"],
                                                        "resources": ["deployments/finalizers"],
                                                        "verbs": ["update"],
                                                    },
                                                    {
                                                        "api_groups": [""],
                                                        "resources": ["pods"],
                                                        "verbs": ["get"],
                                                    },
                                                    {
                                                        "api_groups": ["apps"],
                                                        "resources": [
                                                            "replicasets",
                                                            "deployments",
                                                        ],
                                                        "verbs": ["get"],
                                                    },
                                                    {
                                                        "api_groups": ["pulumi.com"],
                                                        "resources": ["*"],
                                                        "verbs": [
                                                            "create",
                                                            "delete",
                                                            "get",
                                                            "list",
                                                            "patch",
                                                            "update",
                                                            "watch",
                                                        ],
                                                    },
                                                ], opts=pulumi.ResourceOptions(parent=self))

        self.operator_role_binding = kubernetes.rbac.v1.RoleBinding("operatorRoleBinding",
                                                               metadata={
                                                                   "name": "pulumi-kubernetes-operator",
                                                               },
                                                               subjects=[{
                                                                   "kind": "ServiceAccount",
                                                                   "name": "pulumi-kubernetes-operator",
                                                               }],
                                                               role_ref={
                                                                   "kind": "Role",
                                                                   "name": "pulumi-kubernetes-operator",
                                                                   "api_group": "rbac.authorization.k8s.io",
                                                               }, opts=pulumi.ResourceOptions(parent=self))

        self.operator_deployment = kubernetes.apps.v1.Deployment("operatorDeployment",
                                                            metadata={
                                                                "name": "pulumi-kubernetes-operator",
                                                            },
                                                            spec={
                                                                "replicas": 1,
                                                                "selector": {
                                                                    "match_labels": {
                                                                        "name": "pulumi-kubernetes-operator",
                                                                    },
                                                                },
                                                                "template": {
                                                                    "metadata": {
                                                                        "labels": {
                                                                            "name": "pulumi-kubernetes-operator",
                                                                        },
                                                                    },
                                                                    "spec": {
                                                                        "service_account_name": "pulumi-kubernetes-operator",
                                                                        "image_pull_secrets": [{
                                                                            "name": "pulumi-kubernetes-operator",
                                                                        }],
                                                                        "containers": [{
                                                                            "name": "pulumi-kubernetes-operator",
                                                                            "image": "pulumi/pulumi-kubernetes-operator:v0.0.7",
                                                                            "command": ["pulumi-kubernetes-operator"],
                                                                            "args": ["--zap-level=debug"],
                                                                            "image_pull_policy": "Always",
                                                                            "env": [
                                                                                {
                                                                                    "name": "WATCH_NAMESPACE",
                                                                                    "value_from": {
                                                                                        "field_ref": {
                                                                                            "field_path": "metadata.namespace",
                                                                                        },
                                                                                    },
                                                                                },
                                                                                {
                                                                                    "name": "POD_NAME",
                                                                                    "value_from": {
                                                                                        "field_ref": {
                                                                                            "field_path": "metadata.name",
                                                                                        },
                                                                                    },
                                                                                },
                                                                                {
                                                                                    "name": "OPERATOR_NAME",
                                                                                    "value": "pulumi-kubernetes-operator",
                                                                                },
                                                                            ],
                                                                        }],
                                                                    },
                                                                },
                                                            }, opts=pulumi.ResourceOptions(parent=self))
